### PR TITLE
Set field on dataclass transform decorator

### PR DESCRIPTION
**Pull request by NeilGirdhar**
# What does this PR do?

- Set field on dataclass transform decorator so that type checkers know about the field descriptor.
- Use the dataclass transform on the class rather than the metaclass.
- Use `typing_extensions.dataclass_transform`

Fixes # (issue)

## Checklist
- [x] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other
      checks if that's the case).

### CODE REVIEW

**Comment by marcvanzee** [comment]
@NeilGirdhar this PR currently seems to have failing tests, could you please take a look?

**Comment by NeilGirdhar** [comment]
@marcvanzee Thanks for taking a look!  I looked at the error:
```python traceback
File "/home/runner/work/flax/flax/flax/struct.py", line 25, in <module>: Can't find module 'typing_extensions.dataclass_transform'. [import-error]
```
However, this pull request is supposed to install [typing-extensions version 4.1.0](https://github.com/google/flax/pull/1927/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R33), which defines `dataclass_transform` according to [the release notes](https://github.com/google/flax/pull/1927/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R33).

Perhaps it's not being installed during testing for some reason?  How can I check which version is installed during testing?

**Comment by NeilGirdhar** [comment]
Oh, I think I see it, they have fixed some kind of import error in 4.1.1.  Let me try that.

**Comment by marcvanzee** [comment]
Hmm the error is still there. Does it also appear when you run the tests locally? Looking at the build action it seems the `typing` library is installed correctly.

**Comment by NeilGirdhar** [comment]
@marcvanzee Locally, I don't get any errors.

> Looking at the build action it seems the typing library is installed correctly.

What about `typing_extensions`?  Can you verify its version?

**Comment by marcvanzee** [comment]
Ah sorry I actually meant `typing_extensions`, which is installed with version `4.1.1`.

Actually it seems to be a pytype problem, which we also run as part of `run_all_tests.sh`:

```
FAILED: /home/runner/work/flax/flax/.pytype/pyi/flax/struct.pyi 
/opt/hostedtoolcache/Python/3.7.12/x64/bin/python -m pytype.single --imports_info /home/runner/work/flax/flax/.pytype/imports/flax.struct.imports --module-name flax.struct -V 3.7 -o /home/runner/work/flax/flax/.pytype/pyi/flax/struct.pyi --analyze-annotated --nofail --quick /home/runner/work/flax/flax/flax/struct.py
File "/home/runner/work/flax/flax/flax/struct.py", line 25, in <module>: Can't find module 'typing_extensions.dataclass_transform'. [import-error]
```

Do you get the error when you run pytype locally?

**Comment by NeilGirdhar** [comment]
@marcvanzee Yes, I do!  My mistake then.  I'm not sure how to get pytype to see the new typing-extensions?

**Comment by marcvanzee** [comment]
I think you can just ignore the error by appending `# pytype: disable=import-error` to the line 25 in `struct.py`.

**Comment by NeilGirdhar** [comment]
@marcvanzee Okay, I'll do that.  I guess [this](https://github.com/google/pytype/blob/main/pytype/overlays/typing_extensions_overlay.py) needs an update.

**Comment by NeilGirdhar** [comment]
@marcvanzee Woohoo, looks like it passed.

**Review by marcvanzee — APPROVED** [review]
Looks good to me, but we should run internal tests on this to ensure nothing breaks, and I'd like @jheek or @levskaya to take a look at this as well.

**Comment by levskaya** [comment]
Sadly this is causing thousands of internal test failures due to pytype not supporting this mechanism... since we have no option but to use it we may have to delay this PR until pytype properly supports `dataclass_transform`...

**Comment by levskaya** [comment]
Yeah, I think there's still incompatibility with this approach and their existing `dataclass_overlay` extension: https://github.com/google/pytype/blob/main/pytype/overlays/dataclass_overlay.py

**Comment by NeilGirdhar** [comment]
@levskaya I don't understand why there would be any change to the approach?  I just replaced the locally-defined `__dataclass_decorator__` with the imported one from `typing_extensions.dataclass decorator`.  I understand that they don't want to support that yet, but does this introduce any new incompatibilities?

**Comment by levskaya** [comment]
I'm not sure what's wrong exactly, one would have to dive into pytype internals to figure it out, but I see a lot of errors like this:
```
  ...
  File ".../pytype/attribute.py", line 242, in _get_instance_attribute
    return self._get_attribute(node, obj, cls, name, valself)
  File ".../pytype/attribute.py", line 275, in _get_attribute
    node, attr = self._get_member(node, obj, name, valself)
  File ".../pytype/attribute.py", line 485, in _get_member
    obj.load_lazy_attribute(name, subst)
  File ".../pytype/abstract/mixin.py", line 225, in load_lazy_attribute
    variable = self._convert_member(self._member_map[name], subst)
  File ".../pytype/overlay.py", line 41, in _convert_member
    val = member(self.ctx)
  File ".../pytype/overlays/flax_overlay.py", line 39, in make
    return super().make(ctx, "flax.struct")
  File ".../pytype/overlays/dataclass_overlay.py", line 35, in make
    return super().make("dataclass", ctx, mod)
  File ".../pytype/abstract/_pytd_function.py", line 66, in make
    self = cls(name, f.signatures, pyval.kind, ctx)
AttributeError: 'Unsolvable' object has no attribute 'signatures'
```

**Comment by NeilGirdhar** [comment]
@levskaya Right, I assume it will all work when they decide to implement PEP 681.

**Comment by NeilGirdhar** [comment]
@levskaya The Pytype people were kind enough to quickly [fix the issue](https://github.com/google/pytype/issues/1135#issuecomment-1054701517) you ran into.  Hopefully, we can test it out again at their next release.

**Comment by NeilGirdhar** [comment]
@levskaya Would you mind testing it again?   It should work with the new pytype that was released a few minutes ago.

**Comment by levskaya** [comment]
I think this could work w. the new pytype fixes - unfortunately our copy of `typing_extensions` in the internal monorepo is months old and missing the `dataclass_transform` (it has only been in GH typing repo for a month).  I'm trying to see what I need to do to update it, but it touches some core stuff, so I might need to recruit some internal python maintainers to wrangle this.  -- so sorry this is such a pain.

**Comment by NeilGirdhar** [comment]
@levskaya Thanks for your help!

**Comment by levskaya** [comment]
OK, small update - I've updated our internal typing_extensions and it looks like most everything is working... except for some important internal code for which this change is triggering some very odd pytype complaints in a situation with multiple levels of inheritance and classmethod decorators.  I'm trying to figure out how to repro that in isolation so I can figure out how to fix it.   Sorry this is such a slow process... python typechecking is kind of a mess really.

**Comment by NeilGirdhar** [comment]
@levskaya Thanks for the update, I really appreciate your help!

**Comment by levskaya** [comment]
@NeilGirdhar - ok, after a few hours I was able to make a repro of what looks like a pytype bug and reported it to them.  I think once this bug is resolved I can probably try to land this change.  No idea how long that will take :)  Thanks for the patience.

File: flax/linen/module.py [hunk 1]
```diff
@@ -22,6 +22,7 @@
 import typing
 from typing import (Any, Callable, Dict, Iterable, List, Optional,
                     Set, Tuple, Type, TypeVar, Union, overload)
+from typing_extensions import dataclass_transform # pytype: disable=import-error
```
[the remark this answers is not in the document]
Comment (NeilGirdhar) [reply flax/linen/module.py:25]: I'll remove it.  I added that in an effort to remove your problem.  Thanks.

[the remark this answers is not in the document]
Comment (NeilGirdhar) [reply flax/linen/module.py:25]: (Sorry, just sick in bed right now.)

[the remark this answers is not in the document]
Comment (levskaya) [reply flax/linen/module.py:25]: Oh I think something like that might be necessary, it seemed internally that I needed: 
`# pytype: disable=not-supported-yet`
But this is a small detail that we can probably deal with later.
(I hope you get better soon!)

**Comment by levskaya** [comment]
The pytype bug was fixed but will take another ~week to deploy.  I'll try to finally land this change once that's done.  Again, thanks for your patience.

**Comment by NeilGirdhar** [comment]
@levskaya Thank you!

**Comment by levskaya** [comment]
So after that marathon this finally could go in, I also unpinned our pytype version and fixed some more pytype issues.  Do let us know if this helps with outside tooling!

**Pull request merged**